sdks/java/bom/build.gradle

```diff
@@ -0,0 +1,160 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: org.apache.beam.gradle.BeamModulePlugin
+
+apply plugin: "java"
+apply plugin: "maven-publish"
+
+def isRelease(Project project) {
+  return project.hasProperty('isRelease')
+}
+
+ext {
+  mavenJavaDir = "$project.buildDir/publications/mavenJava"
+  mavenJavaBomOutputFile = file(mavenJavaDir + "/pom-default.xml")
+}
+
+for (String projName : rootProject.ext.allProjectNames) {
+  if (!projName.equals(':' + project.name)) {
+    evaluationDependsOn(projName)
+  }
+}
+
+// Copy our pom.xml to the location where a generated POM would go
+task copyPom() {
+  doLast {
+    def moduleNames = new ArrayList<>()
+    for (String projName : rootProject.ext.allProjectNames) {
+      def subproject = project(projName)
+      if (subproject.ext.properties.containsKey('includeInJavaBom') &&
+          subproject.ext.properties.includeInJavaBom) {
+        moduleNames.add(subproject.name)
+      }
+    }
+
+    new File(mavenJavaDir).mkdirs()
+    copy {
+      from 'pom.xml.template'
+      into mavenJavaDir
+      rename 'pom.xml.template', 'pom-default.xml'
+      expand(version: project.version, modules: moduleNames)
+    }
+  }
+}
+
+assemble.dependsOn copyPom
+
+// We want to use our own pom.xml instead of the generated one, so we disable
+// the pom.xml generation and have the publish tasks depend on `copyPom` instead.
+tasks.whenTaskAdded { task ->
+  if (task.name == 'generatePomFileForMavenJavaPublication') {
+    task.enabled = false
+  } else if (task.name == 'publishMavenJavaPublicationToMavenLocal') {
+    task.dependsOn copyPom
+  } else if (task.name == 'publishMavenJavaPublicationToMavenRepository') {
+    task.dependsOn copyPom
+  }
+}
+
+jar.enabled = false
+
+// Remove the default jar archive which is added by the 'java' plugin.
+configurations.archives.artifacts.with { archives ->
+  def artifacts = []
+  archives.each {
+    if (it.file =~ 'jar') {
+      // We can't just call `archives.remove(it)` here because it triggers
+      // a `ConcurrentModificationException`, so we add matching artifacts
+      // to another list, then remove those elements outside of this iteration.
+      artifacts.add(it)
+    }
+  }
+  artifacts.each {
+    archives.remove(it)
+  }
+}
+
+artifacts {
+  archives(mavenJavaBomOutputFile) {
+    builtBy copyPom
+  }
+}
+
+afterEvaluate {
+  // We can't use the `publishing` section from applyJavaNature because
+  // we don't want all the Java artifacts, and we want to use our own pom.xml
+  // instead of the generated one.
+  publishing {
+    publications {
+      mavenJava(MavenPublication) {
```

Comment:
Should consuming this BOM become the recommended way to depend on Beam? If so, perhaps we should implement it in the [maven-archetypes](https://github.com/apache/beam/tree/master/sdks/java/maven-archetypes). The archetype projects also get continuous testing with nightly releases, so that would help ensure our produced BOM is valid.

Reply:
Sounds good; please follow-up in a subsequent PR.